### ISSUE TITLE
Favor composition over inheritance

### ISSUE THREAD

**Issue by rpkamp** [Behat/Behat#1316]
Instead of extending several Symfony classes, we should favour composition instead. For example, for the dispatcher it could be something like:

```
class TestworkEventDispatcher
{
    private $eventDispatcher;

    public function __construct()
    {
        $this->eventDispatcher = new Symfony\Component\Event\EventDispatcher();
    }

    public function addListener($eventName, $event)
    {
        $this->eventDispatcher->addListener($eventName, $event);
    }

    public function dispatch($event)
    {
        $this->dispatcher->dispatch($event);
    }
}
```

This has the advantage that the Symfony Event dispatcher truly becomes an implementation detail, and can be swapped at any moment if it doesn't serve its purpose anymore. Also, it would prevent situations like we just had where a deprecation in the Symfony Dispatcher caused all classes that used the EventDispatcher to change as well.
Had there been encapsulation instead of inheritance the `TestworkEventDispatcher` could have done the BC layer and the rest of the code wouldn't need to be touched.

This is a BC break, since some extensions may depend on the concrete implementation of Symfony. So if we do this there should be a nice upgrade path.

**Comment by ciaranmcnulty** [comment]
We now have a `4.x` branch where we can propose BC breaks

**Comment by stof** [comment]
For the event dispatcher, maybe we should use the PSR-14 interface as type instead of creating our own interface in Behat.

**Comment by acoulton** [comment]
tl;dr - This is too big to be in scope for the 4.0 release in particular because of the impact on extension authors (which will further delay getting real-world upgrades to 4.0). We should definitely consider it for a future major.

That said, there is one dependency we can fairly easily solve - our `Extension` interface should not depend on the symfony `CompilerPassInterface`. This is the first place where we & extension authors are forced into potential BC breaks to support new symfony majors (even if the extension itself never interacts with the DI container). #1796 fixes that so that the `Extension` interface is entirely in our control and we use composition to wire the extension into the DI container.

For the others:

* `symfony/config` - extensions may have a dependency on this if they implement the `Extension::configure()` method to add custom configuration. There's no PSR interface for config-builders, and writing our own generic interface/bridge layer would be fairly complex given the chaining & nesting of calls. It would also be a significant manual job for extension authors to migrate to that.
* `symfony/dependency-injection` - extensions may have a dependency on this if they implement the `Extension::process()` or `Extension::load()` methods to configure their own services, or replace Behat services. Although there is the `PSR-11 Container Interface` this is focused on _using_ a DI container rather than _building_ one. There is for example no support for tagged services, which Behat uses heavily. Again therefore writing our own interfaces / bridge and asking extension authors to migrate to those is a big job.
* `symfony/event-dispatcher` - Is perhaps a bit simpler than the first two. Most extensions don't directly use the event dispatcher, they generally subscribe by registering a service tagged with `EventDispatcherExtension::SUBSCRIBER_TAG` which is expected to implement the symfony `EventSubscriberInterface`. The `EventDispatcherExtension` takes care of finding those & registering them to the dispatcher. We could in theory refactor quite a bit of that towards a more PSR-14 compliant structure (there are some conceptual differences in event objects, and the PSR doesn't entirely cover how listeners should subscribe for events). However, this is another where much of the work would fall on extension authors. I think it'd probably be better (and feasible) to look at this after 4.0 - possibly introducing new subscriber interfaces that work in parallel with the symfony ones first so that extensions (and Behat's internals) can migrate over time, and then eventually dropping support for the symfony interfaces in a future major.

Instead, we should say that for the 3.x and 4.x series, we expect that if extension authors themselves use any of the symfony-based objects we provide, they should define their own `composer.json` constraints on the symfony versions they support. That prevents the Behat major version being a proxy for the symfony version, and will make it easier to safely handle future symfony majors.
